Enable OAuth integration per team/per provider

gokmen opened the pull request:
Each team can allow members to use oauth integration from different providers. This is the main requirement for GitLab integration on multiple teams on one on-premise install of Koding. But with the changes included in this PR it will allow us to keep oauth data per team/per user/per provider.
## Description

Makes oauth integration multiple across teams, providers, users. Updates GitLab oauth integration to use this new oauth mechanism and deprecates existing old way implementations for other oauth providers.
## Motivation and Context

Existing implementation was designed to use oauth integration per user/per provider but with multiple teams on different subdomains we needed to create different applications on providers for each team which can be handled by team admins.
## How Has This Been Tested?

Tested manually for GitLab provider, other tests are wip.
## Screenshots (if appropriate):

![](http://g.recordit.co/ZfJDFBzakK.gif)
![Dashboard App Anchor Support](https://cloud.githubusercontent.com/assets/42368/19127008/18032f78-8af2-11e6-8f39-93e82c10d857.png)
## Types of changes
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
## Checklist:
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
## TODO
- [x] Move `foreignAuth` data into it's own collection from `JUser`
- [x] Keep sensitive information about team oauth integration in another collection
- [x] Provide UI for team admins to set up/disable oauth integrations (GitLab as OAuth provider)
- [x] Check and verify provided oauth endpoint and provided credentials
- [x] Update existing GitLab integration to work with new foreignAuth mechanism
- [x] Disable deprecated oauth handlers
- [x] Make sure only one `foreignId, provider, username, group` on `JForeignAuth`
- [x] Remove existing oauth data on oauth setup change
- [x] Remove sensitive information of oauth integration from `JGroupData` collection on integration disable
- [x] Bongo improvements for authenticateUser flow and not-handled callbacks
- [x] Anchor support for Dashboard sections
- [x] Show onboarding notifications to team members once oauth integration is set-up by team admin

gokmen committed:
isFeatureEnabled: check gitlab feature from active team's config

gokmen committed:
OAuth: provide protocol from backend and update fe part

gokmen committed:
JGroup::fetchData: correct check on provided path with . suffix

gokmen committed:
JUser::extractOauthFromSession -> JSession::fetchOAuthInfo

gokmen committed:
JUser: schema add gitlab into foreignAuth for now

gokmen committed:
JUser::clearOauthFromSession -> JSession::clearOAuthInfo

gokmen committed:
GitLabProvider: check gitlab integration from team config

Enable GitLab for dev environment as well

gokmen committed:
OAuth: pass urlOptions as another option, tidy up the whole flow

gokmen committed:
OAuth: check gitlab config from active team and fetch applicationId for.

gokmen committed:
Code cleanup, tidy up

gokmen committed:
JForeignAuth: initial commit, a model to keep all oauth data per group/member

gokmen committed:
JUser: remove oauth information from JUser model

gokmen committed:
Server: helpers: remove oauth actions on JUser

gokmen committed:
JUser: use JForeignAuth where oauth operation is needed

gokmen committed:
JAccount: use JForeignAuth where oauth operations required

gokmen committed:
OAuth: use custom build hostname for redirectURI

gokmen committed:
Fix linter issues across models

gokmen committed:
ForeignAuth: update persistOauthInfo signature

gokmen committed:
JUser: recaptcha test remove github oauth test for recaptcha

usirin committed:
client/home: add integrations tab

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
client/home: fix test folder structure/requires

sinan commented on workers/social/lib/social/models/user/index.coffee:
```diff
@@ -68,11 +68,6 @@ module.exports = class JUser extends jraphical.Module
-      'foreignAuth.facebook.foreignId' : 'ascending'
-      'foreignAuth.google.foreignId'   : 'ascending'
-      'foreignAuth.linkedin.foreignId' : 'ascending'
-      'foreignAuth.twitter.foreignId'  : 'ascending'
```
👍

sinan reviewed (approved):
looks good just minor things

sinan commented on client/home/lib/index.coffee:
```diff
@@ -19,11 +20,11 @@ module.exports = class HomeAppController extends AppController
     # { title : 'Payment History', viewClass : HomePaymentHistory }
     { title : 'Koding Utilities', viewClass : HomeUtilities, role: 'member' }
     { title : 'My Account', viewClass : HomeAccount, role: 'member' }
+    { title : 'Integrations', viewClass : HomeIntegrations }
```
oh this one is not minor.

gokmen committed:
Remove deprecated integrations helper

gokmen committed:
JAccount: remove deprecated isAuthorized method

gokmen committed:
OAuth: url needs to be fetched from group config, code tidy up

gokmen committed:
Utils: parseClient: simple helper to parse client data where needed

gokmen committed:
JForeignAuth: fetchData: to fetch foreign data for provided session obj.

This will provide the same data that we have before in JUser but this
time this will take group info into account while querying existing
foreign auth data.

gokmen committed:
GitLab_Callback: fetch team settings on oauth register process

gokmen committed:
GitLab_Callback: use provided team settings per request

gokmen committed:
JForeignAuth: code cleanup for create operations

gokmen committed:
JUser: convert: remove deprecated oauth check on convert

gokmen committed:
Ignore outdated go/src/github.com

gokmen committed:
Client: use app/react instead

gokmen committed:
Landing: show sign in with gitlab button if gitlab enabled for team

gokmen committed:
OAuthUrl: use the existing session context for fetching client

gokmen committed:
JForeignAuth: fetchFromSession: update arguments list as options

gokmen committed:
JForeignAuth: fetchFromSession: attach JUser object since it's required

gokmen committed:
OAuth: return error for oauth get url requests when integration is disabled

gokmen committed:
JForeignAuth: fetchData: fix recursive function to return collected data

gokmen committed:
Server: remove deprecated auth-key handlers for old kite implementation

gokmen committed:
JKodingKey: remove deprecated collection definition

gokmen committed:
Package.json: add dns, url-parse and ip packages for internal url checks

gokmen committed:
Helpers: isAddressValid: validity check for provided address

This helper will return error for private IPs and not reachable addresses

gokmen committed:
GitLab: OAuth callback: make sure provided address is valid for oauth

gokmen committed:
Helpers: isAddressValid: no need to lookup on IPv4 addresses

gokmen committed:
JGroup: setOauth for setting up oauth provider per team

>>> a = _remote.revive(_globals.currentGroup);
<<< null

>>> a.setOauth({provider: 'gitlab', enabled: false}, _kd.log)
<<< null

>>> a.setOauth({provider: 'gitlab', enabled: true, applicationId: "5cd0f0f465a91a41", url: "http://10.0.0.1:3000", applicationSecret:"72943dc79eea8"}, _kd.log)
<<< Object {message: "Invalid URL provided", name: "KodingError", isOperational: true}

>>> a.setOauth({provider: 'gitlab', enabled: true, applicationId: "5cd0f0f465a91a41", url: "http://52.52.52.12:3000", applicationSecret:"72943dc79eea8"}, _kd.log)
<<< null

gokmen committed:
Move helpers::isAddress to utils::isAddress for generic usage

gokmen committed:
Test:Logs: skip papertrail tests for now

gokmen committed:
GitLab:OAuth check state of existing session._id to prevent CSRF attacks

usirin committed:
component-lab: fix border-radius for modals

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
component-lab: add ConfirmModal

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
home/integrations: implement GitLab integration UI

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
component-lab: fix content line-height of modal

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

usirin committed:
home/integrations: don't cleanup to default values on cancel

Signed-off-by: Umut Sirin <umutsirin1@gmail.com>

gokmen committed:
OAuth: getUrl: shape up the model based on the PROVIDERS

gokmen committed:
OAuth: clean-up deprecated methods

gokmen committed:
OAuth: use generic errors for common cases

gokmen committed:
OAuth: add ::validateOAuth method for all providers

gokmen committed:
OAuth:validateOAuth with https://tools.ietf.org/html/rfc6749#section-4.4

gokmen committed:
JGroup:setOAuth: use OAuth.validateOAuth method before saving

gokmen committed:
JAccount::updateCounts: remove deprecated updateCounts and it's deps.

gokmen committed:
JGroup: fetchDataAt: share with client side for admins

gokmen committed:
Server: disable deprecated oauth handlers

gokmen committed:
ReviveOAuth: use JForeignAuth for foreignAuth checks/queries

gokmen committed:
GitLabProvider: use provided oauth data of requester

usirin committed:
home/integrations: make sure app secret is being shown

gokmen committed:
ForeignAuth moved into a separate collection

gokmen committed:
GitProviders: GitLab: simply flow by supporting oauth_token

gokmen committed:
GitProvider: importStackTemplateData: use async methods as async only

gokmen committed:
JForeignAuth: keep only one foreign auth info per user/group/provider

gokmen committed:
JGroup: setOAuth: remove sensitive data on oauth disable

gokmen committed:
JGroup: setOAuth: remove existing authentication data on oauth disable

gokmen committed:
Utils: getOAuthEndPoint: returns oauth app endpoint for given provider

gokmen committed:
Integrations: add quick follow up doc for creating integration app

gokmen committed:
OAuth: GitLab: clean provided url before taking any action with it

cihangir commented on workers/social/lib/social/models/foreignauth.coffee:
```diff
@@ -0,0 +1,131 @@
+KodingError = require '../error'
+{ parseClient } = require './utils'
+
+module.exports = class JForeignAuth extends jraphical.Module
```
since we dont use relationships, jraphical.Module no needed, bongo Base or Model is sufficient

gokmen replied:
not `Base` but `Model` 👍

cihangir commented on workers/social/lib/social/models/foreignauth.coffee:
```diff
@@ -0,0 +1,131 @@
+      #
+      #   - { username: 1, group: 1, provider: 1, foreignId: 1 } (unique)
+      #
+      provider    : 'sparse'
```
if we are going to add a compound index, singular sparse indexes are not required

gokmen replied:
I kept it there for development versions since we can validate those indexes for development purposes and currently compound index is just for protection since create flow doesn't allow multiple foreign auths for same user on same team with same provider.

cihangir commented on workers/social/lib/social/models/foreignauth.coffee:
```diff
@@ -0,0 +1,131 @@
+      provider    : 'sparse'
+      foreignId   : 'sparse'
+      group       : 'sparse'
+    sharedEvents  :
```
`sharedEvents ` is not required

cihangir commented on workers/social/lib/social/models/foreignauth.coffee:
```diff
@@ -0,0 +1,131 @@
+        return callback err  if err
+        return callback new KodingError 'User not found'  unless user
+
+        callback err, { user, foreignData }
```
callback null, { user, foreignData }

cihangir commented on workers/social/lib/social/models/gitprovider/gitlabprovider.coffee:
```diff
@@ -27,77 +27,54 @@ injectQueryStrings = (templateData) ->
     port    = if port = gitlab.port then ":#{port}" else ''
-    baseUrl = "http://#{gitlab.host}#{port}" # FIXME update protocol here ~GG
+    # FIXME update protocol here ~GG
+    baseUrl = gitlabHost ? "http://#{gitlab.host}#{port}"
```
it seems we really need to have this in config :)

gokmen replied:
protocol is part of `gitlab.host` now 👍

cihangir commented on workers/social/lib/social/models/gitprovider/gitlabprovider.coffee:
```diff
@@ -130,42 +110,44 @@ module.exports = GitLabProvider =
-          if branches.filter((_branch) -> _branch.name is branch).length is 0
+
+          branches = branches.filter (_branch) ->
+            _branch.name is branch
```
thanks for this version 👍

cihangir commented on workers/social/lib/social/models/group/index.coffee:
```diff
@@ -1143,6 +1155,69 @@ module.exports = class JGroup extends Module
+            return callback err  if err
+
+            dataToUnset = {}
+            dataToUnset["data.#{provider}.applicationSecret"] = null
```
i am not sure if it works with setting to null, but i am sure it works with empty string, if it is working ignore my comment

gokmen replied:
yup, https://docs.mongodb.com/manual/reference/operator/update/unset/#up._S_unset

> The specified value in the $unset expression (i.e. "") does not impact the operation.

gokmen replied:
tested and verified 👍

cihangir commented on workers/social/lib/social/models/group/index.coffee:
```diff
@@ -1143,6 +1155,69 @@ module.exports = class JGroup extends Module
+            dataToUnset["data.#{provider}.applicationSecret"] = null
+            data.update { $unset: dataToUnset }, callback
+
+        return
```
`return` hurts readability here, what about moving if block's content into another function? smth like

`return processNotEnabled <args...> if not enabled`

cihangir commented on workers/social/lib/social/models/group/index.coffee:
```diff
@@ -1143,6 +1155,69 @@ module.exports = class JGroup extends Module
+      validateOptions = { url, applicationId, applicationSecret }
+
+      OAuth.validateOAuth provider, validateOptions, (err) =>
+
```
extraneous space

cihangir commented on workers/social/lib/social/models/oauth.coffee:
```diff
@@ -15,49 +16,171 @@ module.exports = class OAuth extends bongo.Base
+
+        { returnUrl, redirectUri } = urlOptions
+        { applicationId, host, port } = KONFIG.gitlab
         host ?= 'gitlab.com'
```
reading from config here too

cihangir commented on workers/social/lib/social/models/session.coffee:
```diff
@@ -142,10 +142,28 @@ module.exports = class JSession extends Model
+
+
+  @clearOauthInfo = (session, callback) ->
+
```
extraneous space

gokmen commented on servers/lib/server/gitlab_callback.coffee:
```diff
@@ -1,23 +1,63 @@
+
+getPathFor = (url, path) ->
+  { gitlab } = KONFIG
+  url ?= "http://#{gitlab.host ? 'gitlab.com'}:#{gitlab.port ? 80}"
```
updated this one, that `gitlab.com` wasn't necessary at all

gokmen commented on servers/lib/server/index.coffee:
```diff
@@ -120,12 +118,12 @@ app.get  '/-/api/ssotoken/login'                 , require './handlers/api/ssoto
 app.post '/-/api/gitlab', bodyParser.json()      , require './handlers/api/gitlab'
 app.get  '/-/image/cache'                        , require './image_cache'
-app.get  '/-/oauth/github/callback'              , require './github_callback'
+# app.get  '/-/oauth/github/callback'              , require './github_callback'
```
I'm planning to get them back pretty soon, so I kept all the code as is for now. I can do that in the future PR if I won't be able to make it happen as I planned, ok?

gokmen committed:
PR Review fixes for https://github.com/koding/koding/pull/9057

gokmen committed:
JGroup: disableOAuth: fix cyclomatic complexity

cihangir commented:
LGTM, you can merge whenever it is ready.

gokmen committed:
AppStorage: queue up all storage fetchers to reduce calls 7 to 1 on load

gokmen committed:
JAccount: setLastLoginTimezonOffset: deprecated

gokmen committed:
JUser: authenticateClient: make sure every request ends with callback

gokmen committed:
PaymentController: do not call deprecated paymentApi.subscriptions

gokmen committed:
MainApp: use presence endpoint for authentication verification

We're now hitting presence endpoint and making sure the authentication
is also handled by presence checker.

gokmen committed:
bongo-client: remove deprecated package ptal @koding/bongo-client

gokmen committed:
Expressify: on main handler use existing session data for auth requests

Currently for authenticateUser methods we are fetching multiple session
data with the same result. Instead of doing that we're using the already
validated session data before trying to fetch one more time.

This case is only valid for queued requests consists of auth requests.

gokmen committed:
Expressify: process: observe queued requests, send ready to go responses

Currently if any of the requests is failed to call the callback in a
queued requests all of them were failing to reach back to client-side
since dash was preventing to call the final callback. With this change
it will wait up to 7sec. for whole queue to end up and it will return
the results that it got until that time.

Initially it creates a timed out error response for each requests and
once they return a value it will be replaced in responses list. And for
example for a given 3 requests if the second one won't call the callback
the response will be like this;

 - [req1: true, req2: timed out, req3: true]

before this change there were no response at all.

We do also logging these unhanded queues for debugging purposes with;

 - [Expressify:UNHANDLED_QUEUE]

prefix.

gokmen committed:
Bongo:authenticateUser: no need for another event, use callback directly

gokmen committed:
Bongo: handleRequest: make sure callback of a given request called once

There is no check for preventing to call a given method callback's more
than once. This causing some unexpected issues on the application. And
it's preventing us to use async.parallel since it's not allowing this
situation. To prevent that we are keeping the unique callbackId of the
call and checking it's existence before taking any action. Also for
debug purposes this change provides logs with following format;

 - [Bongo:LEAKED_CALL] CONSTRUCTUR_NAME.METHOD_NAME

gokmen committed:
Expressify: use async.parallel instead of dash

gokmen commented:
cc/ @szkl @cihangir again for the latest changes :)

cihangir reviewed (changes requested):
as i said before, this is so unintuitive. at least could you add expected behaviour as comment?

cihangir commented on node_modules_koding/bongo/lib/expressify.coffee:
```diff
@@ -63,15 +63,31 @@ module.exports = (options = {}) ->
+
+          queue.forEach (_req) ->
+            if _req.method is 'authenticateUser'
+              method  = +(Object.keys _req.callbacks)[0]
```
could you add comment here what is `+(Object.keys _req.callbacks)[0]` ?

cihangir commented on node_modules_koding/bongo/lib/expressify.coffee:
```diff
@@ -84,11 +84,11 @@ module.exports = (options = {}) ->
     observeQueue queueId, calls, responses, callback
 
-    dash workQueue, ->
+    async.parallel workQueue, ->
```
❤️

cihangir commented on node_modules_koding/bongo/lib/expressify.coffee:
```diff
@@ -34,27 +34,64 @@ module.exports = (options = {}) ->
+    observeQueue queueId, calls, responses, callback
+
+    dash workQueue, ->
+      callback responses  if OBSERVE_QUEUE[queueId]
```
could we write why do we check `OBSERVE_QUEUE[queueId]` here?

cihangir commented on workers/social/lib/social/models/user/index.coffee:
```diff
@@ -668,11 +669,15 @@ module.exports = class JUser extends jraphical.Module
-        return logout 'error fetching guest account'  if err
+
+        if err
+          return logout 'error fetching guest account', clientId, callback
```
💯

gokmen committed:
TeamSendInvitesFocusStore: deprecated

gokmen committed:
Headerize: provide anchor'ed links on headers for better navigation

gokmen committed:
KodingRouter: keep anchors as is in provided route

gokmen committed:
Home: RouteHandler: fix handleSection flow, reduce rerouting to 1

gokmen committed:
Home: RouteHandler: fix showWelcomeModal flow, reduce rerouting to 1

gokmen committed:
MyTeam: use headerize and sectionize instead

gokmen committed:
Router: use provided anchor info for taking action on Home app nav

gokmen committed:
Account: MyTeam: anchor handlers for specific input focuses

gokmen committed:
Replace Dashboard sub-links to all lower-case

gokmen committed:
showStackEditor: remove deprecated util

gokmen committed:
GitLab: fix typo, print body if there is no connection error as err.

gokmen committed:
Welcome: add GitLab integration setup as welcome step

gokmen committed:
Home: Actions: fix cyclomatic_complexity

gokmen committed:
GitLabProvider: use url-join where needed

gokmen committed:
AppStorage: fetchStorage: simplify queue flow

gokmen committed:
Expressify: comments about queue observer for requests

cihangir commented on node_modules_koding/bongo/lib/expressify.coffee:
```diff
@@ -89,6 +102,13 @@ module.exports = (options = {}) ->
     observeQueue queueId, calls, responses, callback
 
     async.parallel workQueue, ->
+      # If a given queue couldn't completed all items in it in 7sec.
```
👍
